extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,685 @@
+    // valid AMP.
+    this.timerId_ = incrementLoadingAds(this.getWin());
+    return this.adPromise_.then(rendered => {
+      if (!rendered) {
+        // Was not AMP creative so wrap in cross domain iframe.  layoutCallback
+        // has already executed so can do so immediately.
+        this.renderViaIframe_(true);
+      }
+      this.rendered_ = true;
+    }).catch(error => this.promiseErrorHandler_(error));
+  }
+
+  /** @override  */
+  unlayoutCallback() {
+    // Remove creative and reset to allow for creation of new ad.
+    if (!this.layoutMeasureExecuted_) {
+      return true;
+    }
+    // Iframe or shadow root attached as children.  Cannot delete shadowRoot
+    // but creating new one clears.
+    if (this.element.shadowRoot) {
+      this.element.shadowRoot./*REVIEW*/innerHTML = '';
+    } else {
+      while (this.element.firstChild) {
+        removeElement(this.element.firstChild);
+      }
+    }
+
+    this.stylesheets_.forEach(removeElement);
+    this.stylesheets_ = [];
+    this.adPromise_ = null;
+    // Increment promiseId to cause any pending promise to cancel.
+    this.promiseId_++;
+    this.adUrl_ = null;
+    this.rendered_ = false;
+    this.timerId_ = 0;
+    this.layoutMeasureExecuted_ = false;
+    this.intersectionObserver_ = null;
+    return true;
+  }
+
+
+  /**
+   * Gets the Ad URL to send an XHR Request to.  To be implemented
+   * by network.
+   * @return {!Promise<string>}
+   */
+  getAdUrl() {
+    throw new Error('getAdUrl not implemented!');
+  }
+
+  /**
+   * Extracts creative and verification signature (if present) from
+   * XHR response body and header.  To be implemented by network.
+   * @param {!ArrayBuffer} unusedResponseArrayBuffer content as array buffer
+   * @param {!Headers} unusedResponseHeaders Fetch API Headers object (or polyfill
+   *     for it) containing the response headers.
+   * @return {!Promise<{!AdResponseDef}>}
+   */
+  extractCreativeAndSignature(unusedResponseArrayBuffer,
+                              unusedResponseHeaders) {
+    throw new Error('extractCreativeAndSignature not implemented!');
+  }
+
+  /**
+   * @return {boolean} whether environment supports rendering of AMP creatives
+   *    within publisher page via shadow DOM (otherwise will be rendered within)
+   *    cross domain iframe.  If valid AMP creative, will be rendered early.
+   */
+  supportsShadowDom() {
+    return !!window.Element.prototype.createShadowRoot;
+  }
+
+  /**
+   * Callback executed when AMP creative has successfully rendered within the
+   * publisher page via shadow DOM.  To be overridden by network implementations
+   * as needed.
+   */
+  onAmpCreativeShadowDomRender() {}
+
+  /**
+   * Send ad request, extract the creative and signature from the response.
+   * @param {string} adUrl Request URL to send XHR to.
+   * @return {!Promise<?FetchResponse>}
+   * @private
+   */
+  sendXhrRequest_(adUrl) {
+    const xhrInit = {
+      mode: 'cors',
+      method: 'GET',
```

Comment:
Great. Make sure the server also responds with the `Access-Control-Allow-Credentials` header.